paasta_tools/deployd/watchers.py

```diff
@@ -45,6 +55,130 @@ def startup_checker(self, obj):
+            time.sleep(20)
+
+    def get_at_risk_service_instances(self, draining_hosts):
+        marathon_apps = get_marathon_apps()
+        at_risk_tasks = [task for app in marathon_apps for task in app.tasks if task.host in draining_hosts]
+        self.log.info("At risk tasks: {}".format(at_risk_tasks))
+        service_instances = []
+        for task in at_risk_tasks:
+            app_id = task.app_id.strip('/')
+            service, instance, _, __ = deformat_job_id(app_id)
+            service_instances.append(ServiceInstance(service=service,
+                                                     instance=instance,
+                                                     bounce_by=int(time.time()),
+                                                     watcher=self.__class__.__name__,
+                                                     bounce_timers=None))
+        return service_instances
+
+
+class PublicConfigEventHandler(pyinotify.ProcessEvent):
+
+    def my_init(self, filewatcher):
+        self.filewatcher = filewatcher
+        self.public_config = load_system_paasta_config()
+
+    @property
+    def log(self):
+        name = '.'.join([__name__, self.__class__.__name__])
+        return logging.getLogger(name)
+
+    def filter_event(self, event):
+        if event.name.endswith('.json'):
+            return event
+
+    def watch_new_folder(self, event):
+        if event.maskname == 'IN_CREATE|IN_ISDIR':
+            self.filewatcher.wm.add_watch(event.pathname, self.filewatcher.mask, rec=True)
+
+    def process_default(self, event):
+        self.log.debug(event)
+        self.watch_new_folder(event)
+        event = self.filter_event(event)
+        if event:
+            self.log.debug("Public config changed on disk, loading new config")
+            try:
+                new_config = load_system_paasta_config()
+            except ValueError:
+                self.log.error("Couldn't load public config, the JSON is invalid!")
+                return
+            service_instances = []
+            if new_config != self.public_config:
+                self.log.info("Public config has changed, now checking if it affects any services config shas")
+                self.public_config = new_config
+                service_instances = self.get_service_instances_with_changed_id()
+            if service_instances:
+                self.log.info("Found config change affecting {} service instances, "
+                              "now doing a staggered bounce".format(len(service_instances)))
+                service_instances = rate_limit_instances(service_instances,
+                                                         self.public_config.get_deployd_big_bounce_rate(),
+                                                         self.__class__.__name__)
+            for service_instance in service_instances:
+                self.filewatcher.inbox_q.put(service_instance)
+
+    def get_service_instances_with_changed_id(self):
```

Comment:
I hadn't thought of that. The `list_marathon_service_instances` code looks a bit different because it's checking for things that the other watchers cover (e.g. autoscaling)

Given that this will hopefully replace the oracle method I think it's probably not worth trying to refactor them to both share some code.